controllers/cluster_controller.go

```diff
@@ -18,44 +18,99 @@ package controllers
-	"sigs.k8s.io/controller-runtime/pkg/log"
 )
 
 // ClusterReconciler reconciles a Cluster object
 type ClusterReconciler struct {
 	client.Client
+	Log    logr.Logger
 	Scheme *runtime.Scheme
+
+	recorder record.EventRecorder
 }
 
 //+kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters/finalizers,verbs=update
 
-// Reconcile is part of the main kubernetes reconciliation loop which aims to
-// move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the Cluster object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
 func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	log := r.Log.WithValues("cluster", req.NamespacedName)
+
+	// Fetch the Cluster instance.
+	cluster := &capiv1alpha3.Cluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
 
-	// your logic here
+		return ctrl.Result{}, err
+	}
 
-	return ctrl.Result{}, nil
+	return r.reconcile(ctx, cluster, log)
+}
+
+func (r *ClusterReconciler) reconcile(ctx context.Context, cluster *capiv1alpha3.Cluster, log logr.Logger) (ctrl.Result, error) {
+	if _, ok := cluster.Annotations[ignoreClusterDeletion]; ok {
+		log.Info(fmt.Sprintf("Found annotation %s. Cluster %s/%s will be ignored for deletion", ignoreClusterDeletion, cluster.Namespace, cluster.Name))
+		return ctrl.Result{}, nil
+
+	}
+
+	if !cluster.DeletionTimestamp.IsZero() {
+		log.Info(fmt.Sprintf("Deletion for cluster %s/%s is already applied", cluster.Namespace, cluster.Name))
+		return ctrl.Result{}, nil
+	}
+
+	// immediately delete the cluster if defaultTTL has passed
+	if deletionTimeReached(cluster) {
+		err := r.Client.Delete(ctx, cluster, client.PropagationPolicy(metav1.DeletePropagationBackground))
+		if err != nil {
+			log.Error(err, fmt.Sprintf("unable to delete cluster %s/%s", cluster.Namespace, cluster.Name))
+			return requeue(), nil
+		}
+		log.Info(fmt.Sprintf("Cluster %s/%s will be deleted", cluster.Namespace, cluster.Name))
+
+		return ctrl.Result{}, nil
+	}
+
+	// only send marked for deletion event if we still have ~1h before the cluster gets deleted
+	if deletionEventTimeReached(cluster) {
+		log.Info(fmt.Sprintf("Cluster %s/%s is marked for deletion", cluster.Namespace, cluster.Name))
+		r.submitClusterDeletionEvent(cluster, fmt.Sprintf("Cluster %s/%s is marked for deletion", cluster.Namespace, cluster.Name))
```

Comment:
Sorry, I'm a pain in the ass, I know.

nitpick: Feel free to ignore this one, but saying that a cluster is _marked for deletion_ to me sounds like we are adding an annotation or label or doing something with the cluster. But we are not doing anything, really. It's just that it will be removed in 1h. I think logging `Cluster %s%s will be deleted in aprox %s` with the configured TTL would help.

Reply:
No worries I got used to it by Tim 😆 
Good idea, I adjusted the message to send a TTL when we delete the cluster.